{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%config InlineBackend.figure_format='retina'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "\n",
    "from tqdm import tqdm\n",
    "import numpy as np\n",
    "import scipy.spatial\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import anndata\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_output_path = 'hesc_sc_analysis_plots'\n",
    "os.makedirs(plot_output_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c_enrich_guide_df = pd.read_parquet('enrichment_results/class_anno_enrich_guide.parquet')\n",
    "c_enrich_guide_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c_enrich_guide_df['class_anno'] = c_enrich_guide_df['class_anno'].replace(\n",
    "    {'upregulation of cholesterol biosynthesis': 'upregulation of lipid biosynthesis'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the enriched genes in each annotated classes\n",
    "class_anno_enrich_genes_dict = {}\n",
    "for c in np.unique(c_enrich_guide_df['class_anno']):\n",
    "\n",
    "    # Only consider genes with at least 2 enriched guides\n",
    "    enriched_guide_counts = c_enrich_guide_df[\n",
    "        (c_enrich_guide_df['class_anno'] == c)\n",
    "        & (c_enrich_guide_df['log2fc'] > 0) \n",
    "        & (c_enrich_guide_df['positive_count'] > 1)\n",
    "        & (c_enrich_guide_df['perturbed_gene'] != 'non-targeting')\n",
    "    ]['perturbed_gene'].value_counts()\n",
    "    pre_selected_genes = enriched_guide_counts[enriched_guide_counts > 1].index\n",
    "\n",
    "    # Get the significantly with at least 1 significantly enriched guide\n",
    "    enriched_genes = c_enrich_guide_df[\n",
    "        (c_enrich_guide_df['class_anno'] == c)\n",
    "        & (c_enrich_guide_df['log2fc'] > 1) \n",
    "        & (c_enrich_guide_df['positive_count'] > 2)\n",
    "        & (c_enrich_guide_df['pval_adj'] < 0.001)\n",
    "        & (c_enrich_guide_df['perturbed_gene'].isin(pre_selected_genes))\n",
    "    ]['perturbed_gene'].unique()\n",
    "\n",
    "    class_anno_enrich_genes_dict[c] = enriched_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#anno_class_order = np.array(list(class_anno_enrich_genes_dict.keys()))\n",
    "#anno_class_order[np.argsort([len(class_anno_enrich_genes_dict[c]) for c in anno_class_order])]\n",
    "\n",
    "anno_class_order = ['non-targeting enriched', 'pert GNB2L1', 'pert DBR1',\n",
    "       'pert protein neddylation', 'pert mRNA deadenylation',\n",
    "       'pert mRNA-3 processing', 'non-targeting like',\n",
    "       'pert ubiquitin E3 ligase', 'pert RNA methylation',\n",
    "       'pert mTOR signaling', 'germ layer differentiation',\n",
    "       'pert DNA damage checkpoint', 'low mito-genes',\n",
    "       'pert mRNA transcription', 'pert translation', 'low UMI count',\n",
    "       'mesenchymal differentiation', 'pert spliceosome',\n",
    "       'pert cell cycle', 'upregulation of stress response',\n",
    "       'upregulation of lipid biosynthesis']\n",
    "\n",
    "anno_class_sizes = np.array([len(class_anno_enrich_genes_dict[c]) for c in anno_class_order])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(3, 5), dpi=200)\n",
    "\n",
    "ax.barh(anno_class_order, anno_class_sizes, color='grey')\n",
    "ax.set_xlabel('Number of enriched peturbations')\n",
    "ax.set_ylabel('Cell class')\n",
    "\n",
    "fig.savefig(f'{plot_output_path}/perturb_class_enriched_gene_count.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "class_anno_map = {\n",
    "'0_0' : 'non-targeting enriched',\n",
    "'10_0' : 'non-targeting like',\n",
    "'11_0' : 'upregulation of lipid biosynthesis',\n",
    "'12_0' : 'upregulation of stress response',\n",
    "'13_0' : 'non-targeting enriched',\n",
    "'14_0' : 'pert cell cycle',\n",
    "'14_1' : 'pert spliceosome',\n",
    "'14_2' : 'pert mRNA-3 processing',\n",
    "'14_3' : 'pert mRNA transcription',\n",
    "'14_4' : 'pert mRNA transcription',\n",
    "'14_5' : 'pert mRNA transcription',\n",
    "'15_0' : 'germ layer differentiation',\n",
    "'15_1' : 'germ layer differentiation',\n",
    "'15_10' : 'germ layer differentiation',\n",
    "'15_11' : 'germ layer differentiation',\n",
    "'15_12' : 'germ layer differentiation',\n",
    "'15_13' : 'germ layer differentiation',\n",
    "'15_14' : 'mesenchymal differentiation',\n",
    "'15_2' : 'germ layer differentiation',\n",
    "'15_3' : 'germ layer differentiation',\n",
    "'15_4' : 'germ layer differentiation',\n",
    "'15_5' : 'non-targeting enriched',\n",
    "'15_6' : 'germ layer differentiation',\n",
    "'15_7' : 'germ layer differentiation',\n",
    "'15_8' : 'germ layer differentiation',\n",
    "'15_9' : 'germ layer differentiation',\n",
    "'16_0' : 'low UMI count',\n",
    "'16_1' : 'low UMI count',\n",
    "'16_2' : 'low UMI count',\n",
    "'16_3' : 'low UMI count',\n",
    "'16_4' : 'low UMI count',\n",
    "'16_5' : 'low UMI count',\n",
    "'16_6' : 'low UMI count',\n",
    "'16_7' : 'pert DBR1',\n",
    "'17_0' : 'non-targeting enriched',\n",
    "'18_0' : 'pert translation',\n",
    "'18_1' : 'pert mTOR signaling',\n",
    "'18_10' : 'pert translation',\n",
    "'18_11' : 'pert translation',\n",
    "'18_12' : 'pert translation',\n",
    "'18_2' : 'pert translation',\n",
    "'18_3' : 'pert translation',\n",
    "'18_4' : 'pert translation',\n",
    "'18_5' : 'pert mTOR signaling',\n",
    "'18_6' : 'pert translation',\n",
    "'18_7' : 'pert translation',\n",
    "'18_8' : 'pert translation',\n",
    "'18_9' : 'pert translation',\n",
    "'19_0' : 'non-targeting enriched',\n",
    "'19_1' : 'non-targeting like',\n",
    "'1_0' : 'non-targeting enriched',\n",
    "'20_0' : 'pert mRNA transcription',\n",
    "'20_1' : 'pert mRNA transcription',\n",
    "'20_10' : 'pert mRNA deadenylation',\n",
    "'20_11' : 'pert mRNA transcription',\n",
    "'20_2' : 'pert GNB2L1',\n",
    "'20_3' : 'pert mRNA deadenylation',\n",
    "'20_4' : 'pert mRNA transcription',\n",
    "'20_5' : 'pert mRNA transcription',\n",
    "'20_6' : 'pert mRNA transcription',\n",
    "'20_7' : 'pert mRNA transcription',\n",
    "'20_8' : 'pert mRNA transcription',\n",
    "'20_9' : 'pert mRNA transcription',\n",
    "'21_0' : 'non-targeting enriched',\n",
    "'21_1' : 'pert DBR1',\n",
    "'22_0' : 'non-targeting enriched',\n",
    "'23_0' : 'pert ubiquitin E3 ligase',\n",
    "'23_1' : 'pert protein neddylation',\n",
    "'24_0' : 'low mito-genes',\n",
    "'24_1' : 'low mito-genes',\n",
    "'24_2' : 'upregulation of stress response',\n",
    "'25_0' : 'mesenchymal differentiation',\n",
    "'25_1' : 'mesenchymal differentiation',\n",
    "'25_2' : 'mesenchymal differentiation',\n",
    "'25_3' : 'low UMI count',\n",
    "'25_4' : 'mesenchymal differentiation',\n",
    "'26_0' : 'pert DBR1',\n",
    "'27_0' : 'pert RNA methylation',\n",
    "'28_0' : 'pert DNA damage checkpoint',\n",
    "'28_1' : 'pert DNA damage checkpoint',\n",
    "'2_0' : 'non-targeting enriched',\n",
    "'3_0' : 'non-targeting enriched',\n",
    "'4_0' : 'non-targeting enriched',\n",
    "'5_0' : 'non-targeting enriched',\n",
    "'6_0' : 'non-targeting enriched',\n",
    "'7_0' : 'non-targeting enriched',\n",
    "'8_0' : 'non-targeting enriched',\n",
    "'9_0' : 'non-targeting enriched',\n",
    "'9_1' : 'non-targeting enriched',\n",
    "}\n",
    "\n",
    "cluster_annotation_map = {\n",
    "    k : class_anno_map[k] + '_' + k for k in class_anno_map.keys()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c_enrich_gene_df = pd.read_parquet('enrichment_results/l2_c_enrich_gene.parquet')\n",
    "c_enrich_gene_df['cluster_name'] = c_enrich_gene_df['cluster'].map(cluster_annotation_map)\n",
    "c_enrich_gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_genes = np.unique(list(class_anno_enrich_genes_dict['germ layer differentiation']) \n",
    "          + list(class_anno_enrich_genes_dict['mesenchymal differentiation']))\n",
    "\n",
    "selected_clusters = [\n",
    "    'germ layer differentiation_15_0',\n",
    "    'germ layer differentiation_15_12',\n",
    "    'germ layer differentiation_15_9',\n",
    "    'germ layer differentiation_15_10',\n",
    "    'germ layer differentiation_15_11',\n",
    "    'germ layer differentiation_15_7',\n",
    "    'germ layer differentiation_15_1',\n",
    "    'germ layer differentiation_15_2',\n",
    "    'germ layer differentiation_15_8',\n",
    "    'germ layer differentiation_15_13',\n",
    "    'germ layer differentiation_15_3',\n",
    "    'germ layer differentiation_15_6',\n",
    "    'germ layer differentiation_15_4',\n",
    "    'mesenchymal differentiation_15_14',\n",
    "    'mesenchymal differentiation_25_1',\n",
    "    'mesenchymal differentiation_25_0',\n",
    "    'mesenchymal differentiation_25_2',\n",
    "    'mesenchymal differentiation_25_4',\n",
    "]\n",
    "\n",
    "selected_cep_df = c_enrich_gene_df[\n",
    "    c_enrich_gene_df['perturbed_gene'].isin(selected_genes)\n",
    "    & c_enrich_gene_df['cluster_name'].isin(selected_clusters)\n",
    "    ]\n",
    "\n",
    "log2fc_df = selected_cep_df.pivot(index='perturbed_gene', columns='cluster_name', values='log2fc')\n",
    "pval_df = selected_cep_df.pivot(index='perturbed_gene', columns='cluster_name', values='pval_adj')\n",
    "mlog_pval_df = -np.log10(pval_df + 1e-20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pre_ordered_genes = [\n",
    "#    'MED19', 'SP1', 'SUPT20H', 'NANOG', 'SOX2', 'POU5F1',\n",
    "#\n",
    "#    'HARS', 'TARS', 'RRP9', 'GPN3', \n",
    "#\n",
    "#    'CHAF1B', 'MMS22L', 'RNGTT', 'BCL2L1', 'PHB', 'REV3L', 'BRCA2',\n",
    "#    'C22orf15', 'EIF2B4', 'MNAT1', 'TANGO6', 'BRIP1', 'SDE2', 'POLR2M',\n",
    "#    'SMNDC1', 'EIF2B5', 'RNF214', 'PPIE', 'EXOC3', \n",
    "#\n",
    "#    'MRP63', 'RPP14', 'PDCD11', 'DDX56', 'UBE2T',\n",
    "#    'DDX21', 'MED22', 'HEATR1', 'ZC3H8', 'BRIX1',\n",
    "#    'EIF2B3', 'ZNF574', 'SKA3', 'CCNH', 'SKA1', 'IPO7', 'TSEN2',\n",
    "#    'FOXD3', 'EIF2S2', 'PDCD7', 'CENPC', \n",
    "#]\n",
    "#\n",
    "#mlog_pval_df.loc[~mlog_pval_df.index.isin(pre_ordered_genes), :\n",
    "#              ]['mesenchymal differentiation_25_2'].sort_values(ascending=False).index.values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.setdiff1d(selected_genes, gene_order)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gene_order =[\n",
    "#    'MED19', 'SP1', 'SUPT20H', 'NANOG', 'SOX2', 'POU5F1',\n",
    "#\n",
    "#    'HARS', 'TARS', 'RRP9', 'GPN3', \n",
    "#\n",
    "#    'CHAF1B', 'MMS22L', 'RNGTT', 'BCL2L1', 'PHB', 'REV3L', 'BRCA2',\n",
    "#    'C22orf15', 'EIF2B4', 'MNAT1', 'TANGO6', 'BRIP1', 'SDE2', 'POLR2M',\n",
    "#    'SMNDC1', 'EIF2B5', 'RNF214', 'PPIE', 'EXOC3', \n",
    "#\n",
    "#    'MRP63', 'RPP14', 'PDCD11', 'DDX56', 'UBE2T',\n",
    "#    'DDX21', 'MED22', 'HEATR1', 'ZC3H8', 'BRIX1',\n",
    "#    'EIF2B3', 'ZNF574', 'CCNH', 'SKA1', 'IPO7', 'TSEN2',\n",
    "#    'FOXD3', 'EIF2S2', 'PDCD7', 'CENPC', \n",
    "#\n",
    "#    'CENPI', 'RCL1', 'GEMIN5', 'TIMELESS', 'DERL2', 'SHFM1', 'KRR1',\n",
    "#    'RFWD3', 'SKA3', 'TAF6'\n",
    "#]\n",
    "\n",
    "gene_order =[\n",
    "    'MED19', 'SP1', 'SUPT20H', 'NANOG', 'SOX2', 'POU5F1',\n",
    "\n",
    "    'HARS', 'TARS', 'RPP14', 'RPP30', 'TSEN2',\n",
    "\n",
    "    'RRP9', 'PDCD11', 'DDX56', 'DDX21', 'HEATR1', 'BRIX1', 'MRP63', 'RCL1', 'KRR1',\n",
    "\n",
    "    'EIF2B4', 'EIF2B5', 'EIF2S2', 'EIF2B3',\n",
    "\n",
    "    'RNF214', 'UBE2T', 'SHFM1', \n",
    "\n",
    "    'GEMIN5', 'SMNDC1', 'PDCD7', 'PPIE', 'RNGTT', 'EXOSC10',\n",
    "    \n",
    "    'GPN3', 'POLR2M', 'TAF6', 'MED22', 'ZC3H8', 'ZNF574', 'FOXD3',\n",
    "\n",
    "    'SKA1', 'SKA3', \n",
    "    'CENPI', 'CCNH', 'MNAT1', 'CHAF1B', 'TIMELESS',\n",
    "\n",
    "    'MMS22L', 'BRCA2', 'BRIP1', 'SDE2', 'REV3L', 'RFWD3', \n",
    "\n",
    "    'TANGO6', 'EXOC3', 'DERL2',\n",
    "    \n",
    "    'PHB', 'IPO7','BCL2L1','C22orf15', \n",
    "\n",
    "     'MTBP', 'NCAPH', \n",
    "]\n",
    "\n",
    "cluster_order = [\n",
    "   'germ layer differentiation_15_0',\n",
    "    'germ layer differentiation_15_12',\n",
    "    'germ layer differentiation_15_9',\n",
    "    'germ layer differentiation_15_10',\n",
    "    'germ layer differentiation_15_11',\n",
    "    'germ layer differentiation_15_7',\n",
    "    'germ layer differentiation_15_1',\n",
    "    'germ layer differentiation_15_2',\n",
    "    'germ layer differentiation_15_8',\n",
    "    'germ layer differentiation_15_13',\n",
    "    'germ layer differentiation_15_3',\n",
    "    'germ layer differentiation_15_6',\n",
    "    'germ layer differentiation_15_4',\n",
    "    'mesenchymal differentiation_15_14',\n",
    "    'mesenchymal differentiation_25_1',\n",
    "    'mesenchymal differentiation_25_0',\n",
    "    'mesenchymal differentiation_25_2',\n",
    "    'mesenchymal differentiation_25_4',\n",
    "]\n",
    "\n",
    "log2fc_df_to_show = log2fc_df.loc[gene_order, cluster_order].T\n",
    "mlog_pval_df_to_show = mlog_pval_df.loc[gene_order, cluster_order].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "sys.path.append('../')\n",
    "from scattermap import scattermap\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(25, 8))\n",
    "\n",
    "ax = scattermap(log2fc_df_to_show, \n",
    "                marker_size=(7 * mlog_pval_df_to_show.values.astype(float)) + 0.5,\n",
    "                vmin=-6, vmax=6, cmap='coolwarm',\n",
    "                linewidths=0.2, linecolor='black',\n",
    "                ax=ax,\n",
    "                cbar_kws={'shrink':0.5, 'anchor':(0, 0.7)}\n",
    "                )\n",
    "\n",
    "ax.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax.figure.axes[1].tick_params(axis=\"y\", labelsize=7)\n",
    "ax.figure.axes[1].set_ylabel('log2 fold change', fontsize=7)\n",
    "\n",
    "# Create a dot size legend using off-axis scatter calls and legend\n",
    "ax.scatter(-1, -1, label='$10^{-20}$', marker=\"o\", linewidths=0, c=\"grey\", s=140.5)\n",
    "ax.scatter(-1, -1, label='$10^{-10}$', marker=\"o\", linewidths=0, c=\"grey\", s=70.5)\n",
    "ax.scatter(-1, -1, label='$10^{-3}$', marker=\"o\", linewidths=0, c=\"grey\", s=21.5)\n",
    "leg = ax.legend(loc=\"upper left\", bbox_to_anchor=(1, 0.2), fontsize=7)\n",
    "leg.set_title('adjusted p-val',prop={'size':7})\n",
    "\n",
    "fig.savefig(f'{plot_output_path}/perturbed_genes_enriched_in_differentiation_clusters_dot_plot.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mesenchymal_diff_gene_annotation = {\n",
    "    'HARS': 'tRNA synthetase',\n",
    "    'TARS': 'tRNA synthetase',\n",
    "    'RPP14': 'tRNA biogenesis',\n",
    "    'TSEN2': 'tRNA splicing',\n",
    "\n",
    "    'RRP9': 'Ribosome biogenesis',\n",
    "    'PDCD11': 'Ribosome biogenesis',\n",
    "    'DDX56': 'Ribosome biogenesis',\n",
    "    'DDX21': 'Ribosome biogenesis',\n",
    "    'HEATR1': 'Ribosome biogenesis',\n",
    "    'BRIX1': 'Ribosome biogenesis',\n",
    "    'RCL1': 'Ribosome biogenesis',\n",
    "    'KRR1': 'Ribosome biogenesis',\n",
    "    'MRP63' : 'Mitochondrial Ribosomal Protein',\n",
    "\n",
    "    'EIF2B4': 'Translation initiation',\n",
    "    'EIF2B3': 'Translation initiation',\n",
    "    'EIF2B5': 'Translation initiation',\n",
    "    'EIF2S2': 'Translation initiation',\n",
    "\n",
    "    'SKA1': 'chromosome segregation in mitosis', \n",
    "    'SKA3': 'chromosome segregation in mitosis',\n",
    "    'CENPC': 'chromosome segregation in mitosis', \n",
    "    'CENPI': 'chromosome segregation in mitosis',\n",
    "    'CCNH': 'Cyclin H', \n",
    "    'MNAT1' : 'forms the CDK-activating kinase (CAK) enzymatic complex',\n",
    "    'CHAF1B': 'assembly of histone octamers onto newly-replicated DNA',\n",
    "    'TIMELESS': 'Plays an important role in the control of DNA replication',\n",
    "\n",
    "    'SMNDC1' : 'Involved in spliceosome assembly',\n",
    "    'PDCD7': 'a component of the minor U12-type spliceosome',\n",
    "    'PPIE' : 'Involved in pre-mRNA splicing',\n",
    "    'GEMIN5': 'splicing of cellular pre-mRNAs',\n",
    "    'RNGTT' : 'mRNA-capping enzyme',\n",
    "    \n",
    "    'GPN3': 'Small GTPase required for proper localization of RNA polymerase II',\n",
    "    'POLR2M' : 'a subunit of RNA polymerase II',\n",
    "    'TAF6': 'initiation of Pol II transcription',\n",
    "    'MED22': 'component of the Mediator complex',\n",
    "    'ZC3H8': 'transcriptional repressor of the GATA3 promoter',      \n",
    "    'ZNF574': 'May be involved in transcriptional regulation',\n",
    "    'FOXD3': 'TF',\n",
    "\n",
    "    'MMS22L': 'promotes homologous recombination-mediated repair of double-strand breaks', \n",
    "    'BRCA2' : 'double-strand break repair',\n",
    "    'BRIP1' : 'maintenance of chromosomal stability',\n",
    "    'RFWD3': 'response to DNA damage',\n",
    "    'SDE2' : 'cellular response to UV; mitotic G1 DNA damage checkpoint signaling; and protein ubiquitination',\n",
    "    'REV3L' : 'catalytic subunit of DNA polymerase zeta', \n",
    "    \n",
    "    'RNF214' : 'Predicted to enable ubiquitin-protein transferase activity',\n",
    "    'UBE2T': 'Ubiquitin conjugating enzyme',\n",
    "    'SHFM1': 'Component of the 26S proteasome', \n",
    "   \n",
    "    'TANGO6' : 'Predicted to be involved in protein secretion',\n",
    "    'EXOC3': 'docking of exocytic vesicles with fusion sites on the plasma membrane', \n",
    "    'DERL2': 'degradation of misfolded glycoproteins in the ER',\n",
    "\n",
    "    'IPO7': 'nuclear protein import',    \n",
    "    'PHB': 'pleiotropic, mitochondrial chaperone, and transcriptional co-regulator', \n",
    "    'BCL2L1': 'inhibitor of cell death',\n",
    "    'C22orf15': '', \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_genes = [\n",
    "    #'HARS', 'TARS', 'RPP14', 'TSEN2',\n",
    "    #'RRP9', 'PDCD11', 'DDX56', 'DDX21', 'HEATR1', 'BRIX1', 'RCL1', 'KRR1', 'MRP63',\n",
    "    #'EIF2B4', 'EIF2B3', 'EIF2B5', 'EIF2S2',\n",
    "\n",
    "    #'SKA1', 'SKA3', 'CENPC', 'CENPI', 'CCNH', 'MNAT1', 'CHAF1B',\n",
    "\n",
    "    #'SMNDC1', 'PDCD7', 'PPIE', 'GEMIN5', 'RNGTT',\n",
    "\n",
    "    'EIF1AX', 'EIF2B2', 'EIF2S2', 'EIF2S3', 'EIF4A2', 'EIF4E', 'EIF4G1', \n",
    "    'EIF5A', 'EIF5B',\n",
    "    'EIF3J', 'EIF3I', 'EIF3A', 'EIF3C', 'EIF3CL', 'EIF3D', 'EIF3F',\n",
    "    'EIF3M', 'EIF3G', 'EIF4A1', 'EIF3H',\n",
    "    'EIF4G2', 'EIF2B3', 'EIF2B4', 'EIF2B5', 'EIF6'\n",
    "]\n",
    "\n",
    "selected_clusters = [\n",
    "    'upregulation of stress response_12_0', \n",
    "    'pert cell cycle_14_0',\n",
    "    'pert mTOR signaling_18_1',\n",
    "\n",
    "    'pert translation_18_3', \n",
    "    'pert translation_18_0', \n",
    "    'pert translation_18_4',\n",
    "    \n",
    "    'pert translation_18_6', \n",
    "    'pert translation_18_2',\n",
    "       \n",
    "       'pert translation_18_7',\n",
    "    'pert translation_18_12', \n",
    "    \n",
    "       'mesenchymal differentiation_25_0',\n",
    "       'mesenchymal differentiation_25_1'\n",
    "]\n",
    "\n",
    "selected_cep_df = c_enrich_gene_df[\n",
    "    c_enrich_gene_df['perturbed_gene'].isin(selected_genes)\n",
    "    & c_enrich_gene_df['cluster_name'].isin(selected_clusters)\n",
    "    ]\n",
    "\n",
    "log2fc_df = selected_cep_df.pivot(index='perturbed_gene', columns='cluster_name', values='log2fc')\n",
    "pval_df = selected_cep_df.pivot(index='perturbed_gene', columns='cluster_name', values='pval_adj')\n",
    "mlog_pval_df = -np.log10(pval_df + 1e-20)\n",
    "\n",
    "log2fc_df_to_show = log2fc_df.loc[selected_genes, selected_clusters].T\n",
    "mlog_pval_df_to_show = mlog_pval_df.loc[selected_genes, selected_clusters].T\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 4))\n",
    "\n",
    "ax = scattermap(log2fc_df_to_show, \n",
    "                marker_size=(7 * mlog_pval_df_to_show.values.astype(float)) + 0.5,\n",
    "                vmin=-6, vmax=6, cmap='coolwarm',\n",
    "                linewidths=0.2, linecolor='black',\n",
    "                ax=ax,\n",
    "                cbar_kws={'shrink':0.5, 'anchor':(0, 0.7)}\n",
    "                )\n",
    "\n",
    "ax.tick_params(axis='both', which='major', labelsize=7)\n",
    "ax.figure.axes[1].tick_params(axis=\"y\", labelsize=7)\n",
    "ax.figure.axes[1].set_ylabel('log2 fold change', fontsize=7)\n",
    "\n",
    "# Create a dot size legend using off-axis scatter calls and legend\n",
    "ax.scatter(-1, -1, label='$10^{-20}$', marker=\"o\", linewidths=0, c=\"grey\", s=140.5)\n",
    "ax.scatter(-1, -1, label='$10^{-10}$', marker=\"o\", linewidths=0, c=\"grey\", s=70.5)\n",
    "ax.scatter(-1, -1, label='$10^{-3}$', marker=\"o\", linewidths=0, c=\"grey\", s=21.5)\n",
    "leg = ax.legend(loc=\"upper left\", bbox_to_anchor=(1, 0.2), fontsize=7)\n",
    "leg.set_title('adjusted p-val',prop={'size':7})\n",
    "\n",
    "fig.savefig(f'{plot_output_path}/perturbed_eif_genes_enriched_in_differentiation_clusters_dot_plot.pdf', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "c_enrich_gene_df[\n",
    "    c_enrich_gene_df['perturbed_gene'].isin(selected_genes)\n",
    "    & (c_enrich_gene_df['log2fc'] > 1)\n",
    "    & (c_enrich_gene_df['pval_adj'] < 0.01)\n",
    "    & (c_enrich_gene_df['positive_count'] > 1)\n",
    "    ]['cluster_name'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "eif_genes = np.unique(c_enrich_gene_df[c_enrich_gene_df['perturbed_gene'].str.startswith('EIF')]['perturbed_gene'])\n",
    "\n",
    "c_enrich_gene_df[\n",
    "    c_enrich_gene_df['perturbed_gene'].isin(eif_genes)\n",
    "    & (c_enrich_gene_df['log2fc'] > 1)\n",
    "    & (c_enrich_gene_df['pval_adj'] < 0.01)\n",
    "    & (c_enrich_gene_df['positive_count'] > 1)\n",
    "    ]['perturbed_gene'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for g in selected_genes:\n",
    "    for c in class_anno_enrich_genes_dict.keys():\n",
    "        if g in class_anno_enrich_genes_dict[c]:\n",
    "            print(g, c)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scmg",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
