{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff0b6617",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "\n",
    "from tqdm import tqdm\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "import anndata\n",
    "import scanpy as sc\n",
    "import umap\n",
    "import scipy.stats\n",
    "\n",
    "import torch\n",
    "\n",
    "from scmg.model.contrastive_embedding import (CellEmbedder,  embed_adata)\n",
    "\n",
    "from scmg.preprocessing.data_standardization import GeneNameMapper\n",
    "gene_name_mapper = GeneNameMapper()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c66c554b",
   "metadata": {},
   "outputs": [],
   "source": [
    "plot_output_path = 'pathway_rewiring_plots/'\n",
    "os.makedirs(plot_output_path, exist_ok=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a10dc441",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "plt.rcParams[\"figure.autolayout\"] = False\n",
    "matplotlib.rc('pdf', fonttype=42)\n",
    "plt.rcParams['font.family'] = 'FreeSans'\n",
    "sc.set_figure_params(vector_friendly=True, dpi_save=300)\n",
    "plt.rcParams['axes.grid'] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d7f8f36",
   "metadata": {},
   "outputs": [],
   "source": [
    "pert_data_files = [\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/AdamsonWeissman2016_GSM2406681_10X010.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/FrangiehIzar2021_RNA.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/hESC_TF_screen.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/JiangSatija2024_IFNB.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/JiangSatija2024_IFNG.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/JiangSatija2024_INS.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/JiangSatija2024_TGFB.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/JiangSatija2024_TNFA.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/Joung_TFScreen_HS_2023.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/knockTF_human.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/knockTF_mouse.h5ad',\n",
    "#    #'/GPUData_xingjie/SCMG/perturbation_data/omnipath.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/PertOrg.h5ad',\n",
    "    '/GPUData_xingjie/SCMG/perturbation_data/ReplogleWeissman2022_K562_essential.h5ad',\n",
    "    '/GPUData_xingjie/SCMG/perturbation_data/ReplogleWeissman2022_K562_gwps.h5ad',\n",
    "    '/GPUData_xingjie/SCMG/perturbation_data/ReplogleWeissman2022_rpe1.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/TianKampmann2021_CRISPRa.h5ad',\n",
    "#    '/GPUData_xingjie/SCMG/perturbation_data/TianKampmann2021_CRISPRi.h5ad',\n",
    "    '/GPUData_xingjie/SCMG/hESC_perturb_seq/pseudo_bulk_filtered.h5ad', # Test\n",
    "]\n",
    "\n",
    "adata_pert_list = []\n",
    "for pdf in pert_data_files:\n",
    "    adata_pert_list.append(sc.read_h5ad(pdf))\n",
    "    print(os.path.basename(pdf), adata_pert_list[-1].shape[0])\n",
    "\n",
    "adata_pert = anndata.concat(adata_pert_list, axis=0)\n",
    "adata_pert.var['gene_name'] = adata_pert_list[0].var['gene_name']\n",
    "\n",
    "adata_pert"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61e64f67",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mask out the direct target genes\n",
    "for i in range(adata_pert.shape[0]):\n",
    "    pg = adata_pert.obs['perturbed_gene'].iloc[i]\n",
    "    \n",
    "    if pg in adata_pert.var_names:\n",
    "        adata_pert.X[i, adata_pert.var_names.get_loc(pg)] = 0\n",
    "\n",
    "adata_pert.obs['effect_size'] = np.linalg.norm(adata_pert.X, axis=1)\n",
    "adata_pert.obs['max_effect'] = np.max(np.abs(adata_pert.X), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "03a25025",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_exp_cluster_df = pd.read_csv('../../classify_genes/systematic_classification/gene_exp_cluster_annotation.csv', index_col=0)\n",
    "adata_pert.var['gene_exp_cluster'] = gene_exp_cluster_df['cluster_name']\n",
    "gene_exp_cluster_counts = adata_pert.var['gene_exp_cluster'].value_counts()\n",
    "gene_exp_cluster_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5230392d",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array(adata_pert.var[adata_pert.var['gene_exp_cluster'] == 'integrated stress response']['gene_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "224d5202",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_pert.uns['gene_program_names'] = list(gene_exp_cluster_df['cluster_name'].unique())\n",
    "adata_pert.obsm['gene_program'] = np.zeros((adata_pert.shape[0], len(adata_pert.uns['gene_program_names'])))\n",
    "\n",
    "for i in range(len(adata_pert.uns['gene_program_names'])):\n",
    "    genes = list(gene_exp_cluster_df[gene_exp_cluster_df['cluster_name'] == adata_pert.uns['gene_program_names'][i]].index)\n",
    "    if len(genes) == 0:\n",
    "        continue\n",
    "    scale_factor = np.sqrt(len(genes))\n",
    "    adata_pert.obsm['gene_program'][:, i] = scale_factor * np.mean(adata_pert[:, genes].X, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d3ecf64",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_func_emb_df = pd.read_parquet('gene_func_emb_no_hesc.parquet')\n",
    "adata_pert = adata_pert[adata_pert.obs['perturbed_gene_name'].isin(gene_func_emb_df.index)].copy()\n",
    "adata_pert.obsm['gene_func_emb'] = gene_func_emb_df.loc[adata_pert.obs['perturbed_gene_name']].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68c46cfb",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_pert.obs['condition'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7dcdb23",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_k562 = adata_pert[adata_pert.obs['condition'] == 'ReplogleWeissman2022_K562_gwps'].copy()\n",
    "adata_k562_e = adata_pert[adata_pert.obs['condition'] == 'ReplogleWeissman2022_K562_essential'].copy()\n",
    "adata_rpe1 = adata_pert[adata_pert.obs['condition'] == 'ReplogleWeissman2022_rpe1'].copy()\n",
    "adata_hesc = adata_pert[adata_pert.obs['condition'] == 'hESC_perturb_seq'].copy()\n",
    "\n",
    "common_genes = np.intersect1d(adata_k562_e.obs['perturbed_gene'], adata_k562.obs['perturbed_gene'])\n",
    "len(common_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1633c233",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_k562.obs.index = list(adata_k562.obs['perturbed_gene'])\n",
    "adata_k562 = adata_k562[common_genes, :].copy()\n",
    "adata_k562_e.obs.index = list(adata_k562_e.obs['perturbed_gene'])\n",
    "adata_k562_e = adata_k562_e[common_genes, :].copy()\n",
    "\n",
    "adata_rpe1.obs.index = list(adata_rpe1.obs['perturbed_gene'])\n",
    "adata_hesc.obs.index = list(adata_hesc.obs['perturbed_gene'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b336228",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_k562.obs['k562_e_corr'] = [\n",
    "    scipy.stats.pearsonr(\n",
    "        adata_k562.X[i], adata_k562_e.X[i]\n",
    "    )[0] for i in range(adata_k562.shape[0])\n",
    "]\n",
    "\n",
    "adata_k562.obs['k562_e_corr'].hist(bins=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c5f2385",
   "metadata": {},
   "outputs": [],
   "source": [
    "high_k562_intra_corr_genes = adata_k562.obs[adata_k562.obs['k562_e_corr'] > 0.4].index.tolist()\n",
    "print(len(high_k562_intra_corr_genes))\n",
    "\n",
    "adata_k562 = adata_k562[high_k562_intra_corr_genes, :].copy()\n",
    "adata_k562_e = adata_k562_e[high_k562_intra_corr_genes, :].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "567cbb5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "sc.pp.neighbors(adata_k562, n_neighbors=5, use_rep='gene_func_emb', \n",
    "    metric='cosine',\n",
    "    #metric='euclidean',\n",
    ")\n",
    "sc.tl.umap(adata_k562, random_state=0)\n",
    "sc.tl.leiden(adata_k562, flavor=\"igraph\", n_iterations=10, resolution=5, random_state=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7123253c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#adata_k562.write_h5ad('adata_function_cluster.h5ad')\n",
    "adata_k562 = sc.read_h5ad('adata_function_cluster.h5ad')\n",
    "adata_k562"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fb273400",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(4, 4), dpi=200)\n",
    "\n",
    "sc.pl.umap(adata_k562, color='leiden', legend_loc='on data', ax=ax,\n",
    "           legend_fontsize=10, legend_fontoutline=2, s=20,\n",
    "           show=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e53569ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array(adata_k562.obs[adata_k562.obs['leiden'] == '45']['perturbed_gene_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e1d925d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "pg_name_map = {\n",
    "    '0' : 'ribosome large unit biogenesis',\n",
    "    '1' : 'mixed functions',\n",
    "    '2' : 'histone acetylation',\n",
    "    '3' : 'rRNA processing 1',\n",
    "    '4' : 'ribosome large unit protein',\n",
    "    '5' : 'ribosome small unit biogenesis',\n",
    "    '6' : 'proteasome',\n",
    "    '7' : 'chromatin organization',\n",
    "    '8' : 'DNA replication',\n",
    "    '9' : 'mitochondrion organization',\n",
    "    '10' : 'protein neddylation',\n",
    "    '11' : 'mixed functions',\n",
    "    '12' : 'pol II transcription',\n",
    "    '13' : 'translation initiation',\n",
    "    '14' : 'mixed functions',\n",
    "    '15' : 'mixed functions',\n",
    "    '16' : 'mediator complex',\n",
    "    '17' : 'rRNA processing 2',\n",
    "    '18' : 'protein processing in ER',\n",
    "    '19' : 'pol II transcription',\n",
    "    '20' : 'ribosome large unit biogenesis',\n",
    "    '21' : 'DNA replication',\n",
    "    '22' : 'RNA metabolism',\n",
    "    '23' : 'mitochondrial translation',\n",
    "    '24' : 'mixed functions',\n",
    "    '25' : 'ribosome small unit biogenesis',\n",
    "    '26' : 'spliceosome',\n",
    "    '27' : 'transcription initiation',\n",
    "    '28' : 'CCT complex',\n",
    "    '29' : 'spliceosome',\n",
    "    '30' : 'mitochondrion organization',\n",
    "    '31' : 'rRNA processing 2',\n",
    "    '32' : 'Recycling Of eIF2:GDP',\n",
    "    '33' : 'mediator complex',\n",
    "    '34' : 'membrane fission',\n",
    "    '35' : 'histone acetylation',\n",
    "    '36' : 'mitochondrial transcription',\n",
    "    '37' : 'spliceosome',\n",
    "    '38' : 'integrator complex',\n",
    "    '39' : 'ribosome large unit protein',\n",
    "    '40' : 'pol II enlongation',\n",
    "    '41' : 'mRNA surveillance',\n",
    "    '42' : 'RNA degradation',\n",
    "    '43' : 'spliceosome',\n",
    "    '44' : 'rRNA processing 1',\n",
    "    '45' : 'ribosome small unit protein',\n",
    "    '46' : 'RNA methylation',\n",
    "    '47' : 'mitochondrial translation',\n",
    "    '48' : 'spliceosome',\n",
    "    '49' : 'mRNA Polyadenylation',\n",
    "}\n",
    "\n",
    "adata_k562.obs['cluster_name'] = adata_k562.obs['leiden'].map(pg_name_map)\n",
    "adata_k562.write_h5ad('adata_function_cluster_annotated.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f9dc4dc9",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(5, 5), dpi=200)\n",
    "\n",
    "sc.pl.umap(adata_k562, color='cluster_name', \n",
    "        legend_loc='on data', legend_fontsize=8, legend_fontoutline=2, \n",
    "        s=100, ax=ax, title='gene function cluster',\n",
    "         show=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "237bb719",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_anno = adata_k562[adata_k562.obs['cluster_name'] != 'mixed functions'].copy()\n",
    "\n",
    "sc.pp.neighbors(adata_anno, n_neighbors=10, use_rep='gene_func_emb', \n",
    "    metric='cosine',\n",
    "    #metric='correlation',\n",
    "    #metric='euclidean',\n",
    ")\n",
    "sc.tl.umap(adata_anno, random_state=0)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 8), dpi=100)\n",
    "sc.pl.umap(adata_anno, color='cluster_name', legend_loc='on data', ax=ax, \n",
    "           legend_fontsize=10, legend_fontoutline=2)\n",
    "fig.savefig(f'{plot_output_path}/k562_gene_function_clusters_umap.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "495231d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_anno.obs[['perturbed_gene_name', 'cluster_name']].to_csv('perturbation_cluster_annotation.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aae7680b",
   "metadata": {},
   "outputs": [],
   "source": [
    "pg_id_to_cluster_map = dict(zip(adata_anno.obs.index, adata_anno.obs['cluster_name']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "069b2605",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_k562.obs['cluster_name'] = adata_k562.obs.index.map(pg_id_to_cluster_map)\n",
    "k562_group_df = pd.DataFrame(\n",
    "    index=list(adata_anno.obs['cluster_name'].unique()),\n",
    "    columns=adata_k562.uns['gene_program_names'],\n",
    "    dtype=float,\n",
    ")\n",
    "\n",
    "for pg_cluster in k562_group_df.index:\n",
    "    pg_genes = list(adata_k562.obs[adata_k562.obs['cluster_name'] == pg_cluster].index)\n",
    "    k562_group_df.loc[pg_cluster] = adata_k562[pg_genes].obsm['gene_program'].mean(axis=0)\n",
    "\n",
    "\n",
    "adata_rpe1.obs['cluster_name'] = adata_rpe1.obs.index.map(pg_id_to_cluster_map)\n",
    "rpe1_group_df = pd.DataFrame(\n",
    "    index=list(adata_anno.obs['cluster_name'].unique()),\n",
    "    columns=adata_rpe1.uns['gene_program_names'],\n",
    "    dtype=float,\n",
    ")\n",
    "\n",
    "for pg_cluster in rpe1_group_df.index:\n",
    "    pg_genes = list(adata_rpe1.obs[adata_rpe1.obs['cluster_name'] == pg_cluster].index)\n",
    "    rpe1_group_df.loc[pg_cluster] = adata_rpe1[pg_genes].obsm['gene_program'].mean(axis=0)\n",
    "\n",
    "adata_hesc.obs['cluster_name'] = adata_hesc.obs.index.map(pg_id_to_cluster_map)\n",
    "hesc_group_df = pd.DataFrame(\n",
    "    index=list(adata_anno.obs['cluster_name'].unique()),\n",
    "    columns=adata_hesc.uns['gene_program_names'],\n",
    "    dtype=float,\n",
    ")\n",
    "\n",
    "for pg_cluster in hesc_group_df.index:\n",
    "    pg_genes = list(adata_hesc.obs[adata_hesc.obs['cluster_name'] == pg_cluster].index)\n",
    "    hesc_group_df.loc[pg_cluster] = adata_hesc[pg_genes].obsm['gene_program'].mean(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb4ce702",
   "metadata": {},
   "outputs": [],
   "source": [
    "pg_cluster_order = [\n",
    "    'chromatin organization', 'DNA replication', 'histone acetylation', 'CCT complex', \n",
    "    'mediator complex', 'transcription initiation', 'integrator complex', 'pol II transcription', 'pol II enlongation', \n",
    "    'spliceosome', 'RNA metabolism', 'RNA degradation', 'RNA methylation', 'mRNA Polyadenylation', 'mRNA surveillance', \n",
    "\n",
    "\n",
    "    'ribosome small unit biogenesis', \n",
    "    'ribosome small unit protein', 'rRNA processing 1', 'rRNA processing 2', \n",
    "    'ribosome large unit biogenesis', 'ribosome large unit protein',\n",
    "    'translation initiation', 'Recycling Of eIF2:GDP', \n",
    "\n",
    "    'protein neddylation', 'proteasome',\n",
    "    'protein processing in ER', 'membrane fission',\n",
    "\n",
    "       'mitochondrion organization', \n",
    "       \n",
    "       \n",
    "       'mitochondrial translation', \n",
    "       \n",
    "       'mitochondrial transcription', \n",
    "       ]\n",
    "\n",
    "gene_program_order = [\n",
    "    'chromatin structure', 'DNA replication/repair', 'cell cycle (G1/S)', 'cell cycle (prometaphase)', 'cell cycle (M phase)', \n",
    "    'spliceosome', 'proliferation',\n",
    "    'ribosome biogenesis', 'ribosomal protein genes', 'mitochondrial encoded', 'cholesterol biosynthesis', 'mitochondrial ribosome', \n",
    "    'Golgi vesicle transport', 'unfolded protein response', \n",
    "    'integrated stress response', 'p53 signaling', 'lysosome/autophagy',  \n",
    "\n",
    "       \n",
    "#    'pluripotency', 'Hox genes', 'glia', 'neural development', 'neuronal', 'peripheral neurons', \n",
    "#    'visual perception', 'retinal epithelium', 'melanin biosynthesis', \n",
    "#    'epithelial', 'respiratory epithelium', 'kidney', 'intestine', 'pancreatic', 'pancreatic islet', 'liver', 'epidermal', \n",
    "#    'mesothelial', 'adipocyte',  'mesenchymal', 'smooth muscle', 'endothelial', 'bone', \n",
    "#\n",
    "#    'interferon signaling', 'TNF signaling', 'immune system', 'myeloid', 'macrophage', 'B cell', 'T cell', 'natural killer',\n",
    "#    'mast cell', 'erythroid', 'megakaryocyte',\n",
    "#       \n",
    "#    'muscle', 'heart', 'cilia', \n",
    "       \n",
    "    ]\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "sns.heatmap(k562_group_df.loc[pg_cluster_order, gene_program_order],\n",
    "            cmap='seismic', center=0, cbar_kws={'label': 'mean gene program shift'},\n",
    "            #xticklabels=False, yticklabels=False,\n",
    "            vmax=2, vmin=-2,\n",
    "            ax=ax\n",
    "            )\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "ax.set_xlabel('Gene expression programs')\n",
    "ax.set_ylabel('Functional clusters')\n",
    "ax.set_title('K562')\n",
    "fig.savefig(f'{plot_output_path}/k562_gene_program_shift_heatmap.pdf')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "sns.heatmap(rpe1_group_df.loc[pg_cluster_order, gene_program_order],\n",
    "            cmap='seismic', center=0, cbar_kws={'label': 'mean gene program shift'},\n",
    "            #xticklabels=False, yticklabels=False,\n",
    "            vmax=2, vmin=-2,\n",
    "            ax=ax\n",
    "            )\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "ax.set_xlabel('Gene expression programs')\n",
    "ax.set_ylabel('Functional clusters')\n",
    "ax.set_title('RPE1')\n",
    "fig.savefig(f'{plot_output_path}/rpe1_gene_program_shift_heatmap.pdf')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(8, 8))\n",
    "sns.heatmap(hesc_group_df.loc[pg_cluster_order, gene_program_order],\n",
    "            cmap='seismic', center=0, cbar_kws={'label': 'mean gene program shift'},\n",
    "            #xticklabels=False, yticklabels=False,\n",
    "            vmax=1, vmin=-1,\n",
    "            ax=ax\n",
    "            )\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "ax.set_xlabel('Gene expression programs')\n",
    "ax.set_ylabel('Functional clusters')\n",
    "ax.set_title('hESC')\n",
    "fig.savefig(f'{plot_output_path}/hesc_gene_program_shift_heatmap.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "320f6af4",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_to_show = [\n",
    "    'ATF5', 'DDIT3', 'GDF15', 'XBP1', 'HSPA5', 'PDIA4', 'ATF4', 'ATF3', 'GARS1', 'EPRS1', 'SARS1', 'WARS1', 'MYC'\n",
    "]\n",
    "\n",
    "genes_to_show = [g for g in genes_to_show if g in list(adata_k562.var['gene_name'])]\n",
    "\n",
    "adata_k562.obs['cluster_name'] = adata_k562.obs.index.map(pg_id_to_cluster_map)\n",
    "k562_pathway_gene_df = pd.DataFrame(\n",
    "    index=list(adata_anno.obs['cluster_name'].unique()),\n",
    "    columns=genes_to_show,\n",
    "    dtype=float,\n",
    ")\n",
    "\n",
    "for pg_cluster in k562_pathway_gene_df.index:\n",
    "    pg_genes = list(adata_k562.obs[adata_k562.obs['cluster_name'] == pg_cluster].index)\n",
    "    gene_indices = [list(adata_k562.var['gene_name']).index(gene) for gene in genes_to_show]\n",
    "\n",
    "    k562_pathway_gene_df.loc[pg_cluster] = adata_k562[pg_genes].X[:, gene_indices].mean(axis=0)\n",
    "\n",
    "\n",
    "adata_rpe1.obs['cluster_name'] = adata_rpe1.obs.index.map(pg_id_to_cluster_map)\n",
    "rpe1_pathway_gene_df = pd.DataFrame(\n",
    "    index=list(adata_anno.obs['cluster_name'].unique()),\n",
    "    columns=genes_to_show,\n",
    "    dtype=float,\n",
    ")\n",
    "\n",
    "for pg_cluster in rpe1_pathway_gene_df.index:\n",
    "    pg_genes = list(adata_rpe1.obs[adata_rpe1.obs['cluster_name'] == pg_cluster].index)\n",
    "    gene_indices = [list(adata_rpe1.var['gene_name']).index(gene) for gene in genes_to_show]\n",
    "\n",
    "    rpe1_pathway_gene_df.loc[pg_cluster] = adata_rpe1[pg_genes].X[:, gene_indices].mean(axis=0) \n",
    "\n",
    "\n",
    "adata_hesc.obs['cluster_name'] = adata_hesc.obs.index.map(pg_id_to_cluster_map)\n",
    "hesc_pathway_gene_df = pd.DataFrame(\n",
    "    index=list(adata_anno.obs['cluster_name'].unique()),\n",
    "    columns=genes_to_show,\n",
    "    dtype=float,\n",
    ")\n",
    "\n",
    "for pg_cluster in hesc_pathway_gene_df.index:\n",
    "    pg_genes = list(adata_hesc.obs[adata_hesc.obs['cluster_name'] == pg_cluster].index)\n",
    "    gene_indices = [list(adata_hesc.var['gene_name']).index(gene) for gene in genes_to_show]\n",
    "\n",
    "    hesc_pathway_gene_df.loc[pg_cluster] = adata_hesc[pg_genes].X[:, gene_indices].mean(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c92d17b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(15, 8))\n",
    "sns.heatmap(k562_pathway_gene_df.loc[pg_cluster_order],\n",
    "            cmap='seismic', center=0, cbar_kws={'label': 'mean gene program shift'},\n",
    "            #xticklabels=False, yticklabels=False,\n",
    "            vmax=0.5, vmin=-0.5,\n",
    "            ax=ax\n",
    "            )\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "ax.set_xlabel('Gene expression programs')\n",
    "ax.set_ylabel('Functional clusters')\n",
    "ax.set_title('K562')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(15, 8))\n",
    "sns.heatmap(rpe1_pathway_gene_df.loc[pg_cluster_order],\n",
    "            cmap='seismic', center=0, cbar_kws={'label': 'mean gene program shift'},\n",
    "            #xticklabels=False, yticklabels=False,\n",
    "            vmax=0.5, vmin=-0.5,\n",
    "            ax=ax\n",
    "            )\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "ax.set_xlabel('Gene expression programs')\n",
    "ax.set_ylabel('Functional clusters')\n",
    "ax.set_title('RPE1')\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(15, 8))\n",
    "sns.heatmap(hesc_pathway_gene_df.loc[pg_cluster_order],\n",
    "            cmap='seismic', center=0, cbar_kws={'label': 'mean gene program shift'},\n",
    "            #xticklabels=False, yticklabels=False,\n",
    "            vmax=0.5, vmin=-0.5,\n",
    "            ax=ax\n",
    "            )\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "ax.set_xlabel('Gene expression programs')\n",
    "ax.set_ylabel('Functional clusters')\n",
    "ax.set_title('hESC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ceb6a75",
   "metadata": {},
   "outputs": [],
   "source": [
    "pathway_of_interest = 'cytoplasmic translation 1'\n",
    "\n",
    "#pathway_of_interest = 'transcription initiation'\n",
    "#pathway_of_interest = 'translation initiation'\n",
    "#pathway_of_interest = 'mitochondrial transcription'\n",
    "\n",
    "gene_list1 = np.array(adata_k562.obs[adata_k562.obs['cluster_name'] == pathway_of_interest]['perturbed_gene_name'])\n",
    "gene_list1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1cd78c55",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list2 = np.array(adata_rpe1.obs[adata_rpe1.obs['cluster_name'] == pathway_of_interest]['perturbed_gene_name'])\n",
    "gene_list2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9739dcb9",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list3 = np.array(adata_hesc.obs[adata_hesc.obs['cluster_name'] == pathway_of_interest]['perturbed_gene_name'])\n",
    "gene_list3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f31f689",
   "metadata": {},
   "outputs": [],
   "source": [
    "common_genes = np.intersect1d(gene_list1, gene_list2)\n",
    "common_genes = np.intersect1d(common_genes, gene_list3)\n",
    "common_genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec945a4a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes_to_show = [\n",
    "    'INO80B', 'NELFA', 'NELFB', 'NELFCD', 'NELFE', 'TAF2', 'TAF6', \n",
    "    'DPH2', 'EIF1AX', 'EIF3M', 'EIF4E', 'EIF4G1', 'STRAP', 'XRCC5',\n",
    "    'AHCY', 'LRPPRC', 'MTPAP', 'POLRMT', 'REXO2', 'SSBP1', 'TEFM', 'TFAM'\n",
    "    ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "202439a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_pert.obs['condition'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "101fddf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_to_show = 'ReplogleWeissman2022_K562_gwps'\n",
    "#genes_to_show = common_genes\n",
    "\n",
    "adata_to_show = adata_pert[adata_pert.obs['condition'] == dataset_to_show]\n",
    "adata_to_show = adata_to_show[adata_to_show.obs['perturbed_gene_name'].isin(genes_to_show)].copy()\n",
    "\n",
    "\n",
    "\n",
    "gene_program_show_df = pd.DataFrame(\n",
    "    index=adata_to_show.obs['perturbed_gene_name'],\n",
    "    columns=adata_pert.uns['gene_program_names'],\n",
    "    data=adata_to_show.obsm['gene_program']\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 6))\n",
    "sns.heatmap(gene_program_show_df.loc[genes_to_show, gene_program_order], cmap='seismic', ax=ax, center=0, vmax=4, vmin=-4,\n",
    "            cbar_kws={'label': 'RMS'})\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "\n",
    "\n",
    "ax.set_xlabel('Gene expression programs')\n",
    "ax.set_ylabel('Perturbed genes')\n",
    "ax.set_title(dataset_to_show)\n",
    "fig.savefig(f'{plot_output_path}/{dataset_to_show}_example_genes_heatmap.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e6cd986e",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_to_show = 'ReplogleWeissman2022_rpe1'\n",
    "#genes_to_show = common_genes\n",
    "\n",
    "adata_to_show = adata_pert[adata_pert.obs['condition'] == dataset_to_show]\n",
    "adata_to_show = adata_to_show[adata_to_show.obs['perturbed_gene_name'].isin(genes_to_show)].copy()\n",
    "\n",
    "gene_program_show_df = pd.DataFrame(\n",
    "    index=adata_to_show.obs['perturbed_gene_name'],\n",
    "    columns=adata_pert.uns['gene_program_names'],\n",
    "    data=adata_to_show.obsm['gene_program']\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 6))\n",
    "sns.heatmap(gene_program_show_df.loc[genes_to_show, gene_program_order], cmap='seismic', ax=ax, center=0, vmax=4, vmin=-4,\n",
    "            cbar_kws={'label': 'RMS'})\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "\n",
    "ax.set_xlabel('Gene expression programs')\n",
    "ax.set_ylabel('Perturbed genes')\n",
    "ax.set_title(dataset_to_show)\n",
    "fig.savefig(f'{plot_output_path}/{dataset_to_show}_example_genes_heatmap.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b565f3c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_to_show = 'hESC_perturb_seq'\n",
    "#genes_to_show = common_genes\n",
    "\n",
    "adata_to_show = adata_pert[adata_pert.obs['condition'] == dataset_to_show]\n",
    "adata_to_show = adata_to_show[adata_to_show.obs['perturbed_gene_name'].isin(genes_to_show)].copy()\n",
    "\n",
    "\n",
    "gene_program_show_df = pd.DataFrame(\n",
    "    index=adata_to_show.obs['perturbed_gene_name'],\n",
    "    columns=adata_pert.uns['gene_program_names'],\n",
    "    data=adata_to_show.obsm['gene_program']\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 6))\n",
    "sns.heatmap(gene_program_show_df.loc[genes_to_show, gene_program_order], cmap='seismic', ax=ax, center=0, vmax=2, vmin=-2,\n",
    "            cbar_kws={'label': 'RMS'})\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "\n",
    "ax.set_xlabel('Gene expression programs')\n",
    "ax.set_ylabel('Perturbed genes')\n",
    "ax.set_title(dataset_to_show)\n",
    "fig.savefig(f'{plot_output_path}/{dataset_to_show}_example_genes_heatmap.pdf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2023cd66",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dataset_to_show = 'ReplogleWeissman2022_K562_gwps'\n",
    "#dataset_to_show = 'ReplogleWeissman2022_rpe1'\n",
    "dataset_to_show = 'hESC_perturb_seq'\n",
    "\n",
    "\n",
    "adata_to_show = adata_pert[adata_pert.obs['condition'] == dataset_to_show]\n",
    "adata_to_show = adata_to_show[adata_to_show.obs['perturbed_gene_name'].str.startswith('EIF')].copy()\n",
    "\n",
    "\n",
    "gene_program_show_df = pd.DataFrame(\n",
    "    index=adata_to_show.obs['perturbed_gene_name'],\n",
    "    columns=adata_pert.uns['gene_program_names'],\n",
    "    data=adata_to_show.obsm['gene_program']\n",
    ")\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(6, 10))\n",
    "sns.heatmap(gene_program_show_df.loc[:, gene_program_order].sort_index(), cmap='seismic', ax=ax, center=0, vmax=2, vmin=-2,\n",
    "            cbar_kws={'label': 'RMS'})\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "\n",
    "ax.set_xlabel('Gene Expression Programs')\n",
    "ax.set_ylabel('Perturbed Genes')\n",
    "ax.set_title(dataset_to_show)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db89a1bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array(sorted(gene_program_show_df.index))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71b60994",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_pert"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a75259f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "pg_to_show = [\n",
    "    'TADA3', 'YEATS2',\n",
    "    'TFAM', 'TEFM', 'POLRMT',\n",
    "]\n",
    "\n",
    "readout_gene_to_show = [\n",
    "    'BAX', 'MDM2', 'ELOB', \n",
    "    'ASPM', 'HMMR', 'TPX2', \n",
    "    'ATF3', 'ATF4', 'ATF5', 'DDIT3',\n",
    "    'GARS1', 'IARS1', 'LARS1',\n",
    "    'INSIG1', 'MVD', 'FDPS', \n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d4716290",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scmg",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
