{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy.stats\n",
    "from tqdm.auto import tqdm\n",
    "\n",
    "import anndata\n",
    "import scanpy as sc\n",
    "\n",
    "from cytofuture_data.gene_name_mapping import GeneNameMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the gene name mapper\n",
    "gene_name_mapper = GeneNameMapper(\n",
    "    '../../standard_genes/gene_names/human_genes.csv',\n",
    "    '../../standard_genes/gene_names/mouse_genes.csv',\n",
    "    '../../standard_genes/gene_names/orthologue_map_human2mouse_best.csv',\n",
    "    '../../standard_genes/gene_names/orthologue_map_mouse2human_best.csv'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = '/home/xingjie/Data/data2/cytofuture/datasets/perturb-seq'\n",
    "dataset_id = 'JiangSatija2024_INS'\n",
    "adata = sc.read_h5ad(os.path.join('/home/xingjie/Data/data2/cytofuture/datasets/perturb-seq',\n",
    "                                  'raw_data', 'JiangSatija2024', 'Seurat_object_INS_Perturb_seq.h5ad'))\n",
    "adata = adata.raw.to_adata()\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "adata.obs['condition'] = [r[1]['cell_type'] + '_' + r[1]['gene'] for r in adata.obs.iterrows()]\n",
    "\n",
    "# Set the gene IDs to Ensembl IDs\n",
    "adata.var['ensembl_id'] = gene_name_mapper.map_gene_names(adata.var.index, 'human', 'human', 'name', 'id')\n",
    "adata.var.index = adata.var['ensembl_id']\n",
    "adata = adata[:, adata.var.index.str.startswith('ENSG')].copy()\n",
    "\n",
    "sc.pp.normalize_total(adata, target_sum=1e4)\n",
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Print the conditions\n",
    "condition_col = 'condition'\n",
    "condition_cell_counts = adata.obs[condition_col].value_counts()\n",
    "for c in condition_cell_counts.index:\n",
    "    print(f'{c}\\t{condition_cell_counts.loc[c]}')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the data frame of perturbations\n",
    "n_cell_threshold = 50\n",
    "perturbation_dict = {\n",
    "    'condition' : [],\n",
    "    'perturbed_gene' : [],\n",
    "    'control_condition': [],\n",
    "}\n",
    "\n",
    "for c in condition_cell_counts.index:\n",
    "    if condition_cell_counts.loc[c] < n_cell_threshold:\n",
    "        continue\n",
    "    \n",
    "    perturbed_genes = gene_name_mapper.map_gene_names([c.split('_')[1]], \n",
    "                                'human', 'human', 'name', 'id')\n",
    "    perturbed_genes = np.intersect1d(perturbed_genes, adata.var.index)\n",
    "    \n",
    "    # Only consider single gene perturbations\n",
    "    if len(perturbed_genes) != 1:\n",
    "        continue\n",
    "    pg = perturbed_genes[0]\n",
    "    control_condition = c.split('_')[0] + '_NT'\n",
    "\n",
    "    # Ignore condtions with too few cells\n",
    "    n_cells = int(condition_cell_counts.loc[c])\n",
    "    if n_cells < n_cell_threshold:\n",
    "        continue\n",
    "\n",
    "    control_exps = adata[adata.obs[condition_col] == control_condition, pg\n",
    "                         ].X.toarray().reshape(-1)\n",
    "    perturbed_exps = adata[adata.obs[condition_col] == c, pg\n",
    "                           ].X.toarray().reshape(-1)\n",
    "    \n",
    "    # Calculate the fold-change and p-val for the perturbed gene\n",
    "    fc = np.mean(perturbed_exps) / np.mean(control_exps)\n",
    "    pval = scipy.stats.mannwhitneyu(control_exps, perturbed_exps, alternative='two-sided')[1]\n",
    "\n",
    "    # Only keep the perturbations that are effective\n",
    "    if (fc < 0.8) and (pval < 1e-2):\n",
    "        perturbation_dict['condition'].append(c)\n",
    "        perturbation_dict['perturbed_gene'].append(pg)\n",
    "        perturbation_dict['control_condition'].append(control_condition)\n",
    "        \n",
    "        print(f'{c}\\t{n_cells}\\t{pg}\\t{fc:.2f}\\t{pval:.2e}')\n",
    "\n",
    "perturbation_df = pd.DataFrame(perturbation_dict).set_index('condition')\n",
    "perturbation_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_cluster_mean_expression_matrix(adata, cluster_column):\n",
    "    '''Get a dataframe of mean gene expression of each cluster.'''\n",
    "    cell_exp_mtx = pd.DataFrame(adata.X.toarray(), index=adata.obs[cluster_column], columns=adata.var.index)    \n",
    "    cluster_mean_exp = cell_exp_mtx.groupby(by=cluster_column).mean()\n",
    "    return cluster_mean_exp\n",
    "\n",
    "def get_cluster_mean_expression_matrix_low_mem(adata, cluster_column):\n",
    "    '''Get a dataframe of mean gene expression of each cluster.'''\n",
    "    cluster_names = np.unique(adata.obs[cluster_column].values)\n",
    "    cluster_mean_df = pd.DataFrame(np.zeros((len(cluster_names), adata.shape[1]), dtype=np.float32), \n",
    "                                   index=cluster_names, columns=adata.var.index)\n",
    "    \n",
    "    for c in tqdm(cluster_names):\n",
    "        X_c = adata[adata.obs[cluster_column] == c].X\n",
    "        cluster_mean_df.loc[c] = X_c.mean(axis=0)\n",
    "    \n",
    "    return cluster_mean_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get mean expressions of each condition\n",
    "selected_conditions = np.unique(list(perturbation_df.index) \n",
    "                                + list(perturbation_df['control_condition']))\n",
    "adata_selected = adata[adata.obs[condition_col].isin(selected_conditions)]\n",
    "\n",
    "condition_mean_exp_df = get_cluster_mean_expression_matrix_low_mem(adata_selected, \n",
    "                                                                   condition_col)\n",
    "condition_mean_exp_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "standard_gene_df = pd.read_csv('standard_genes.csv').set_index('human_id')\n",
    "\n",
    "obs_dict = {\n",
    "    'id' : [],\n",
    "    'condition' : [],\n",
    "    'perturbed_gene' : [],\n",
    "    'perturbation_sign' : [],\n",
    "}\n",
    "\n",
    "X_perturb = []\n",
    "X_control = []\n",
    "X_measure_mask = []\n",
    "\n",
    "# Get the perturbation data\n",
    "for c in tqdm(perturbation_df.index):\n",
    "    obs_dict['id'].append(f'{dataset_id}_{c}')\n",
    "    obs_dict['condition'].append(dataset_id)\n",
    "    obs_dict['perturbed_gene'].append(perturbation_df.loc[c, 'perturbed_gene'])\n",
    "    obs_dict['perturbation_sign'].append(-1) # -1 for knockdown, 1 for overexpression\n",
    "    \n",
    "    control_c = perturbation_df.loc[c, 'control_condition']\n",
    "\n",
    "    standard_gene_df['perturb'] = condition_mean_exp_df.loc[c]\n",
    "    standard_gene_df['control'] = condition_mean_exp_df.loc[control_c]\n",
    "    standard_gene_df['measure_mask'] = ~standard_gene_df['perturb'].isna()\n",
    "\n",
    "    X_perturb.append(np.nan_to_num(standard_gene_df['perturb'].values, \n",
    "                           nan=0).astype(np.float32))\n",
    "    X_control.append(np.nan_to_num(standard_gene_df['control'].values, \n",
    "                           nan=0).astype(np.float32))\n",
    "    X_measure_mask.append(standard_gene_df['measure_mask'].values.astype(np.float32))\n",
    "\n",
    "X_perturb = np.array(X_perturb)\n",
    "X_control = np.array(X_control)\n",
    "X_perturb = np.log1p(X_perturb / X_perturb.sum(axis=1, keepdims=True) * 1e4)\n",
    "X_control = np.log1p(X_control / X_control.sum(axis=1, keepdims=True) * 1e4)\n",
    "X_shift = X_perturb - X_control\n",
    "\n",
    "X_measure_mask = np.array(X_measure_mask)\n",
    "\n",
    "# Create the perturbation anndata\n",
    "adata_perturb = anndata.AnnData(\n",
    "    X=X_shift, \n",
    "    obs=pd.DataFrame(obs_dict).set_index('id'),\n",
    "    var=standard_gene_df[[]].copy(),\n",
    ")\n",
    "adata_perturb.layers['measure_mask'] = X_measure_mask\n",
    "adata_perturb.layers['control'] = X_control\n",
    "\n",
    "adata_perturb.obs['perturbed_gene_name'] = gene_name_mapper.map_gene_names(\n",
    "    adata_perturb.obs['perturbed_gene'], 'human', 'human', 'id', 'name')\n",
    "adata_perturb.var['gene_name'] = gene_name_mapper.map_gene_names(\n",
    "    adata_perturb.var.index, 'human', 'human', 'id', 'name')\n",
    "\n",
    "adata_perturb.write_h5ad(os.path.join(\n",
    "    '/home/xingjie/Data/data2/cytofuture/datasets/perturbation_data', \n",
    "    f'{dataset_id}.h5ad'))\n",
    "\n",
    "plt.hist(adata_perturb.X.flatten(), bins=100)\n",
    "adata_perturb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cytofuture",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
