{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import json\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import scipy.stats\n",
    "from tqdm import tqdm\n",
    "\n",
    "import anndata\n",
    "import scanpy as sc\n",
    "\n",
    "from cytofuture_data.gene_name_mapping import GeneNameMapper"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the gene name mapper\n",
    "gene_name_mapper = GeneNameMapper(\n",
    "    '../../standard_genes/gene_names/human_genes.csv',\n",
    "    '../../standard_genes/gene_names/mouse_genes.csv',\n",
    "    '../../standard_genes/gene_names/orthologue_map_human2mouse_best.csv',\n",
    "    '../../standard_genes/gene_names/orthologue_map_mouse2human_best.csv'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a map from Entrez ID to Ensembl ID\n",
    "entrez_name_df = pd.read_csv(\n",
    "    '/home/xingjie/Data/data2/cytofuture/databases/PertOrg/mart_export_ensembl_entrez.csv')\n",
    "entrez_name_df = entrez_name_df[~entrez_name_df['NCBI gene (formerly Entrezgene) ID'].isna()].copy()\n",
    "entrez_name_df = entrez_name_df.drop_duplicates(subset='NCBI gene (formerly Entrezgene) ID').copy()\n",
    "display(entrez_name_df)\n",
    "\n",
    "entrez_to_ensembl_map = {int(r[1]['NCBI gene (formerly Entrezgene) ID']) : r[1]['Gene stable ID']\n",
    "                        for r in entrez_name_df.iterrows()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the meta data for the database\n",
    "database_df = pd.read_csv(\n",
    "    '/home/xingjie/Data/data2/cytofuture/databases/PertOrg/Genetical_perurbation_Datasets.csv')\n",
    "\n",
    "database_df = database_df[database_df['organism'] == 'Mus_musculus']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the meta data for the database\n",
    "database_df = pd.read_csv(\n",
    "    '/home/xingjie/Data/data2/cytofuture/databases/PertOrg/Genetical_perurbation_Datasets.csv')\n",
    "\n",
    "database_df = database_df[database_df['organism'] == 'Mus_musculus']\n",
    "database_df = database_df[database_df['edittype'].isin([\n",
    "    'knockout', 'transgene', 'overexpression', 'knockdown', 'knockin',\n",
    "])]\n",
    "database_df = database_df.set_index('Pertorgid')\n",
    "database_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "standard_gene_df = pd.read_csv('standard_genes.csv').set_index('human_id')\n",
    "pert_sign_map = {'knockout' : -1, 'transgene' : 1, \n",
    "                 'overexpression' : 1, 'knockdown' : -1, \n",
    "                 'knockin' : 1,}\n",
    "\n",
    "obs_dict = {\n",
    "    'id' : [],\n",
    "    'condition' : [],\n",
    "    'perturbed_gene' : [],\n",
    "    'perturbation_sign' : [],\n",
    "}\n",
    "X_perturb = []\n",
    "X_control = []\n",
    "X_measure_mask = []\n",
    "\n",
    "dataset_id = 'PertOrg'\n",
    "\n",
    "# Get the perturbation data\n",
    "for pertorgid in tqdm(database_df.index):\n",
    "    # Get the perturbed gene\n",
    "    pg = gene_name_mapper.map_gene_names([database_df.loc[pertorgid, 'genotype']],\n",
    "                                         'mouse', 'human', 'name', 'id')[0]\n",
    "    if pg == 'na':\n",
    "        continue\n",
    "\n",
    "    # Load the dataset\n",
    "    dataset_file = os.path.join('/home/xingjie/Data/data2/cytofuture/databases/PertOrg/Differential_Gene_Expression_Analysis_Result_of_Each_PertOrg_Dataset',\n",
    "                    f'{pertorgid}.txt')\n",
    "    \n",
    "    if not os.path.exists(dataset_file):\n",
    "        continue\n",
    "    \n",
    "    dataset_df = pd.read_csv(dataset_file, sep='\\t')\n",
    "    dataset_df = dataset_df[~dataset_df['GeneID'].isna()]\n",
    "    dataset_df['target_gene'] = gene_name_mapper.map_gene_names(\n",
    "                                        dataset_df['GeneID'].astype(int).map(entrez_to_ensembl_map),\n",
    "                                        'mouse', 'human', 'id', 'id')\n",
    "    dataset_df = dataset_df[dataset_df['target_gene'] != 'na']\n",
    "    dataset_df['AveExpr_Control'] = dataset_df['AveExpr_Control'].astype(float)\n",
    "    dataset_df['AveExpr_Case'] = dataset_df['AveExpr_Case'].astype(float)\n",
    "    dataset_df = dataset_df[~dataset_df['AveExpr_Control'].isna()]\n",
    "    dataset_df = dataset_df[~dataset_df['AveExpr_Case'].isna()].copy()\n",
    "    dataset_df = dataset_df.set_index('target_gene')\n",
    "    dataset_df = dataset_df[~dataset_df.index.duplicated(keep='first')]\n",
    "\n",
    "    # Filter out the dataset if there are too few genes \n",
    "    # or the expression counts are not raw\n",
    "    if ((dataset_df.shape[0] < 5000)\n",
    "        or (dataset_df['AveExpr_Case'].min() < 0)\n",
    "        or (dataset_df['AveExpr_Control'].min() < 0)\n",
    "        or (dataset_df['AveExpr_Case'].max() < 100)\n",
    "        or (dataset_df['AveExpr_Control'].max() < 100)):\n",
    "        continue\n",
    "\n",
    "    # Get the sign of perturbation\n",
    "    pert_type = database_df.loc[pertorgid, 'edittype']\n",
    "    perturbation_sign = pert_sign_map[pert_type]\n",
    "\n",
    "    obs_dict['id'].append(f'{dataset_id}_{pertorgid}')\n",
    "    obs_dict['condition'].append(f'{dataset_id}_{pertorgid}')\n",
    "    obs_dict['perturbed_gene'].append(pg)\n",
    "    obs_dict['perturbation_sign'].append(perturbation_sign) # -1 for knockdown, 1 for overexpression\n",
    "    \n",
    "    standard_gene_df['perturb'] = dataset_df['AveExpr_Case']\n",
    "    standard_gene_df['control'] = dataset_df['AveExpr_Control']\n",
    "    standard_gene_df['measure_mask'] = ~standard_gene_df['perturb'].isna()\n",
    "\n",
    "    X_perturb.append(np.nan_to_num(standard_gene_df['perturb'].values, \n",
    "                           nan=0).astype(np.float32))\n",
    "    X_control.append(np.nan_to_num(standard_gene_df['control'].values, \n",
    "                           nan=0).astype(np.float32))\n",
    "    X_measure_mask.append(standard_gene_df['measure_mask'].values.astype(np.float32))\n",
    "\n",
    "X_perturb = np.array(X_perturb)\n",
    "X_control = np.array(X_control)\n",
    "X_perturb = np.log1p(X_perturb / X_perturb.sum(axis=1, keepdims=True) * 1e4)\n",
    "X_control = np.log1p(X_control / X_control.sum(axis=1, keepdims=True) * 1e4)\n",
    "X_shift = X_perturb - X_control\n",
    "\n",
    "X_measure_mask = np.array(X_measure_mask)\n",
    "\n",
    "# Create the perturbation anndata\n",
    "adata_perturb = anndata.AnnData(\n",
    "    X=X_shift, \n",
    "    obs=pd.DataFrame(obs_dict).set_index('id'),\n",
    "    var=standard_gene_df[[]].copy(),\n",
    ")\n",
    "adata_perturb.layers['measure_mask'] = X_measure_mask\n",
    "adata_perturb.layers['control'] = X_control\n",
    "\n",
    "adata_perturb.obs['perturbed_gene_name'] = gene_name_mapper.map_gene_names(\n",
    "    adata_perturb.obs['perturbed_gene'], 'human', 'human', 'id', 'name')\n",
    "adata_perturb.var['gene_name'] = gene_name_mapper.map_gene_names(\n",
    "    adata_perturb.var.index, 'human', 'human', 'id', 'name')\n",
    "\n",
    "adata_perturb.write_h5ad(os.path.join(\n",
    "    '/home/xingjie/Data/data2/cytofuture/datasets/perturbation_data', \n",
    "    f'{dataset_id}.h5ad'))\n",
    "\n",
    "plt.hist(adata_perturb.X.flatten(), bins=100)\n",
    "adata_perturb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.hist(adata_perturb.X.flatten(), bins=100)\n",
    "plt.ylim(0, 1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cytofuture",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
